Add --mapc and for symmetry -mapc

tarsius opened the pull request:
I often wish I could use `--map` instead of `dolist` or `mapc` but when I do then the byte-compiler rightfully complains that I shouldn't be using `mapcar` when throwing away the value. So here's `--mapc` and for symmetry `-mapc` too.

Fuco1 commented:
Dash has `--each` which is like `mapc` with some extra features: like providing `it-index`.  I think better solution then would be to alias `-mapc` to `-each`?

Fuco1 commented:
Granted, `mapc` is faster than `-each` because it's C while `-each` is built on top of a while loop.  Hmm... Do you think it is worth to optimize here?

tarsius commented:
I don't really intend to use `-mapc`, I'll just use `mapc` or `-each` as appropriate. So the performance difference does not really matter to me, but I tend to think this should not be an alias.

Making `-mapc` an alias for `-each` complicates documentation. Currently we have e.g.:

```
#### -map-when `(pred rep list)`
...
Alias: `-replace-where`
```

Meaning `-replace-where` is an alias for `-map-when` **and** `--replace-where` is an alias for `--map-when`. But while `-mapc` could be an alias for `-each`, `--mapc` isn't an alias for `--each`, so things would start falling apart documentation wise.

tarsius commented:
Or are you saying that not even `--mapc` is necessarily useful because there's already `--each`? In that case, I would argue that the better performance of  `--mapc` would indeed be desirable and justify its addition.

Fuco1 commented:
That's what I ment yes, to alias both `-mapc` and `--mapc` to respective versions of `each`.

tarsius commented:
Actually that's not gonna work because the order of the arguments is different.

(I just pushed an amended commit, which puts the documentation in the right place).

Fuco1 commented:
Oh, I see, you're right.

I guess each was mostly ment to be used in the `--each` variant in a way similar to (fictional) `with-list-items`, that is, the first argument declared special and the rest forms magically populated with data (cf. `with-current-buffer`, `with-temp-file` etc.)

Fuco1 commented:
I think I'm okey with merging this, let's wait a couple days if @magnars shows up.

But you should add some description to the docstring as to why/how this is different to `-each` since they now duplicate behaviour and differ only in implementation (`-mapc` being somewhat faster as it is C basd).  Also a "see also" link could be added.

In other news, this patch breaks some compilation tests... and since compilation in emacs is something I totally ignore, I don't know how to fix it.

magnars commented:
How is it different than `-each`? Maybe it could be an alias? And if this one is faster, maybe `-each` could get a speed improvement?

fbergroth commented:
Regarding the failed test; the old test was broken and #113 hopefully fixes it.

Fuco1 commented:
@magnars `mapc` iteration is implemented in C, whereas `-each` uses a while loop. That allows for some additional features like `it-index` which are otherwise (without using lexical scoping) impossible to achieve using just `mapc` and function hacking. They also have swapped arguments.

I don't know how much faster it is though, it would be cool to have some benchmarks.

fbergroth commented:
If it's just a matter of performance, we could scan `body` argument in `--each` for occurrences of `it-index`, and generate code using `mapc` if none are found. Also `it-index` seems to be undocumented in `--each`.

magnars commented:
Sounds like there is room for both. Would you mind rebasing on top of master, @tarsius - to run the tests again with the fixes?

tarsius commented:
I have rebased and updated the documentation, and have pushed the result.

When I run `./run-tests.sh` as-is I get:

```
dev/ert.el: `flet' is an obsolete macro (as of 24.3); use either `cl-flet' or `cl-letf'.
dev/ert.el: `labels' is an obsolete macro (as of 24.3); use `cl-labels' instead.
dev/ert.el: `flet' is an obsolete macro (as of 24.3); use either `cl-flet' or `cl-letf'.
Eager macro-expansion failure: (wrong-type-argument . "match-form must be a list")
Running 131 tests (2015-01-13 18:38:53+0100)
   passed    1/131  !cdr
...
   passed  131/131  -zip-with

Ran 131 tests, 131 results as expected (2015-01-13 18:38:54+0100)

Wrote /home/jonas/.emacs.d/lib/dash/dash.elc
```

When I remove `-l dev/ert.el` which shouldn't be required on `24.4` it get's worse:

```
Eager macro-expansion failure: (wrong-type-argument . "match-form must be a list")
Wrong type argument
```

That's it and the exit code is 255.

I have no idea where the eager macro-expansion failure is coming from and whether it's the cause of the error in the second case, and if so why it's not an error when using the outdated `ert`.

fbergroth commented:
@tarsius The eager macroexpansion failure seems to come from [this test](https://github.com/magnars/dash.el/blob/26c1bbe371db3cff89d4915488a55e381ee51fe5/dev/examples.el#L841).
I'm not sure how to rewrite it though, or if we can just remove it.

Fuco1 commented:
I fear we might end up with "PHP-like" mess with virtually equivalent functions with different argument orders... it might get very frustrating for the users.

tarsius commented:
I am perfectly fine with this not being merged.

magnars commented:
That's okay. Thanks!
